File: benchmarks/ts-bench/src/tests/connection-storm.ts
import type { Config } from '../config.js';
import type { TierResult } from '../report.js';
import * as protocol from '../protocol.js';
import { LatencyHistogram, printHeader } from '../stats.js';
import { generateBenchToken } from '../jwt.js';
import { tiersFor } from '../config.js';

export async function run(config: Config): Promise<TierResult[]> {
  printHeader(
    'TEST 1: Connection Storm',
    'All connections opened as fast as possible. Measures accept rate + handshake latency.'
  );

  const tiers = tiersFor(config, 'connection-storm');
  const token = generateBenchToken(Buffer.from(config.secret), 'bench-user');
  const results: TierResult[] = [];

  for (const n of tiers) {
    console.log(`\n  --- ${n} connections ---`);

    let errors = 0;
    const allLatencies: number[] = [];
    const connections: protocol.WsConnection[] = [];
    const totalStart = process.hrtime.bigint();

    const useMultiIp = n > 60_000 && config.host === '127.0.0.1';
    let connIdx = 0;

    for (let start = 0; start < n; start += config.batchSize) {
      const end = Math.min(start + config.batchSize, n);
      const chunk = end - start;

      const promises: Promise<{ ws: protocol.WsConnection; latMs: number }>[] = [];

      for (let i = 0; i < chunk; i++) {
        const t0 = process.hrtime.bigint();

        let connectPromise: Promise<protocol.WsConnection>;
        if (useMultiIp) {
          const ipIdx = Math.floor(connIdx / 60_000) + 1;
          if (ipIdx > 254) throw new Error('exceeded 127.0.0.254');
          connIdx++;
          connectPromise = protocol.connectAndHandshakeFrom(
            config.host, config.port, token,
            'compression=false&protocol_version=1',
            `127.0.0.${ipIdx}`, 15
          );
        } else {
          connIdx++;
          connectPromise = protocol.connectAndHandshake(
            config.host, config.port, token,
            'compression=false&protocol_version=1', 15
          );
        }

        promises.push(
          connectPromise.then((ws) => {
            const latMs = Number(process.hrtime.bigint() - t0) / 1_000_000;
            return { ws, latMs };
          })
        );
      }

      const settled = await Promise.allSettled(promises);
      for (const r of settled) {
        if (r.status === 'fulfilled') {
          connections.push(r.value.ws);
          allLatencies.push(r.value.latMs);
        } else {
          errors++;
        }
      }

      process.stderr.write(`\r    Progress: ${connections.length}/${n}...`);

      if (end < n) {
        await new Promise((r) => setTimeout(r, 10));
      }
    }

    const totalSecs = Number(process.hrtime.bigint() - totalStart) / 1_000_000_000;

    process.stderr.write(
      `\r    Connected: ${connections.length}/${n} (${errors} failed)            \n`
    );
    console.log(`    Total time: ${totalSecs.toFixed(2)}s`);

    const hist = new LatencyHistogram();
    for (const lat of allLatencies) {
      hist.recordMs(lat);
    }

    if (allLatencies.length > 0) {
      const rate = allLatencies.length / totalSecs;
      console.log(`    Accept rate: ${Math.round(rate)} conn/s`);
      hist.printSummary('    ');
    }

    results.push({
      tier: n,
      connected: connections.length,
      errors,
      durationSecs: totalSecs,
      messagesSent: 0,
      messagesReceived: 0,
      bytesSent: 0,
      latency: hist.isEmpty() ? null : hist.toSummary(),
      extra: {
        accept_rate: totalSecs > 0 ? allLatencies.length / totalSecs : 0,
      },
    });

    await protocol.closeAll(connections);
    console.log();
  }

  return results;
}

File: benchmarks/rust-bench/src/tests/connection_storm.rs
use crate::config::Cli;
use crate::protocol;
use crate::report::{LatencySummary, TierResult};
use crate::stats::{self, LatencyHistogram};
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::Arc;
use std::time::Instant;

/// Test 1: Connection Storm
/// Opens N connections as fast as possible, measuring accept rate and handshake latency.
pub async fn run(cli: &Cli) -> Vec<TierResult> {
    stats::print_header(
        "TEST 1: Connection Storm",
        "All connections opened as fast as possible. Measures accept rate + handshake latency.",
    );

    let tiers = cli.tiers_for(crate::config::TestName::ConnectionStorm);
    let token = crate::jwt::generate_bench_token(cli.secret.as_bytes(), "bench-user");
    let mut results = Vec::new();

    for &n in &tiers {
        println!("\n  --- {} connections ---", n);

        let errors = Arc::new(AtomicUsize::new(0));
        let total_start = Instant::now();

        // Spawn all connection attempts in batches
        let batch_size = cli.batch_size;
        let mut all_latencies: Vec<f64> = Vec::with_capacity(n);
        let mut connections = Vec::with_capacity(n);

        let use_multi_ip = n > 60_000 && cli.host == "127.0.0.1";
        let mut conn_idx = 0usize;

        for start in (0..n).step_by(batch_size) {
            let end = (start + batch_size).min(n);
            let chunk = end - start;

            let mut handles = Vec::with_capacity(chunk);
            for _ in 0..chunk {
                let host = cli.host.clone();
                let port = cli.port;
                let tok = token.clone();
                let err_count = errors.clone();

                if use_multi_ip {
                    let ip_idx = (conn_idx / 60_000) + 1;
                    assert!(
                        ip_idx <= 254,
                        "exceeded 127.0.0.254 — too many connections for loopback"
                    );
                    let source_addr: std::net::SocketAddr =
                        format!("127.0.0.{}:0", ip_idx).parse().unwrap();
                    conn_idx += 1;
                    handles.push(tokio::spawn(async move {
                        let t0 = Instant::now();
                        match protocol::connect_and_handshake_from(
                            &host,
                            port,
                            &tok,
                            "compression=false&protocol_version=1",
                            source_addr,
                            15,
                        )
                        .await
                        {
                            Ok(ws) => {
                                let lat_ms = t0.elapsed().as_secs_f64() * 1000.0;
                                Ok((ws, lat_ms))
                            }
                            Err(_) => {
                                err_count.fetch_add(1, Ordering::Relaxed);
                                Err(())
                            }
                        }
                    }));
                } else {
                    conn_idx += 1;
                    handles.push(tokio::spawn(async move {
                        let t0 = Instant::now();
                        match protocol::connect_and_handshake(
                            &host,
                            port,
                            &tok,
                            "compression=false&protocol_version=1",
                            15,
                        )
                        .await
                        {
                            Ok(ws) => {
                                let lat_ms = t0.elapsed().as_secs_f64() * 1000.0;
                                Ok((ws, lat_ms))
                            }
                            Err(_) => {
                                err_count.fetch_add(1, Ordering::Relaxed);
                                Err(())
                            }
                        }
                    }));
                }
            }

            let batch_results = futures_util::future::join_all(handles).await;
            for r in batch_results {
                if let Ok(Ok((ws, lat))) = r {
                    connections.push(ws);
                    all_latencies.push(lat);
                }
            }

            eprint!("\r    Progress: {}/{}...", connections.len(), n);

            if end < n {
                tokio::time::sleep(std::time::Duration::from_millis(10)).await;
            }
        }

        let total_secs = total_start.elapsed().as_secs_f64();
        let err_count = errors.load(Ordering::Relaxed);

        eprintln!(
            "\r    Connected: {}/{} ({} failed)            ",
            connections.len(),
            n,
            err_count
        );
        println!("    Total time: {:.2}s", total_secs);

        // Build histogram
        let mut hist = LatencyHistogram::new();
        for &lat in &all_latencies {
            hist.record_ms(lat);
        }

        if !all_latencies.is_empty() {
            let rate = all_latencies.len() as f64 / total_secs;
            println!("    Accept rate: {:.0} conn/s", rate);
            hist.print_summary("    ");
        }

        results.push(TierResult {
            tier: n,
            connected: connections.len(),
            errors: err_count,
            duration_secs: total_secs,
            messages_sent: 0,
            messages_received: 0,
            bytes_sent: 0,
            latency: Some(LatencySummary::from_histogram(&hist)),
            extra: serde_json::json!({
                "accept_rate": if total_secs > 0.0 { all_latencies.len() as f64 / total_secs } else { 0.0 }
            }),
        });

        // Cleanup
        protocol::close_all(connections).await;
        println!();
    }

    results
}
